### ISSUE TITLE
Integrate our mingw-w64-git into MSYS2 proper

### ISSUE THREAD

**Issue by dscho** [git-for-windows/git#284]
This ticket entails several tasks.

The biggest one is to get our msys2-runtime patches integrated into upstream (maybe not as-are, but with modifications).

We also need to make sure that [our `mintty` changes](https://github.com/git-for-windows/MSYS2-packages/compare/eaa7852%5E...eaa7852%5E2) are integrated into MSys2, either as-are (into MSYS2-packages), or even better as a new mintty release once https://github.com/mintty/mintty/pull/471 is merged.

The remaining task is to rename our MINGW-packages' `master` branch and restart a new `master` from upstream, just integrating the diff to our `mingw-w64-git`.

**Comment by fabio-porcedda** [comment]
It would be great to be able to install msys2 and have also avaiable the latest version of git for windows!

Thanks for the effort.

**Comment by dscho** [comment]
> It would be great to be able to install msys2 and have also avaiable the latest version of git for windows!
> 
> Thanks for the effort.

I am glad we are in agreement! This is exactly what this ticket is about.

**Comment by maoueh** [comment]
@fabio-porcedda This is already possible if you want. I'm doing this since a few months without problems.

I added instructions on the wiki: https://github.com/git-for-windows/git/wiki/Install-inside-MSYS2-proper

**Comment by nalla** [comment]
@maoueh the steps described there is roughly the same steps the _Git for Windows SDK_ is doing. In fact. the _Git for Windows SDK_ is roughly just that: *MSYS2* with a special repository before the core repositories.

**Comment by maoueh** [comment]
I'm not sure I understand the meaning of your comment. Was it just informational?

**Comment by nalla** [comment]
> Was it just informational
Sure.

**Comment by dscho** [comment]
So here are the Pull Requests I opened so far:

* Alexpux/Cygwin#3
* Alexpux/Cygwin#4
* Alexpux/Cygwin#5
* Alexpux/Cygwin#6
* Alexpux/Cygwin#7
* Alexpux/Cygwin#8
* mintty/mintty#471

Once those are in, I need to open another Pull Request to update the `mintty` MSys2 package, and then another one to switch `mingw-w64-git` to Git for Windows' code base, taking over maintainership [as requested](https://github.com/git-for-windows/MINGW-packages/commit/8d473014ba56441b5612109a35ba48d186a32881#commitcomment-10953357) by @mingwandroid

**Comment by dscho** [comment]
Alexpux/Cygwin#3 was closed because the patch had made it into Cygwin proper in the meantime. The other `msys2-runtime` PRs were updated to the newest `msys2-master`.

**Comment by hansfn** [comment]
This is very interesting. I have been waiting for an easy way to get Git inside MSYS / MinGW Shell for a long time, and now it seems to be very close to happening for MSYS2 - great!

What is the status of this work? It seems also Alexpux/Cygwin#4 was closed because the patch has made it into Cygwin proper, but the other pull requests are hanging.

Thx a lot for your work!

**Comment by dscho** [comment]
> What is the status of this work?

The main problem now is that the following Pull Requests need more or less work (I might have forgotten other things):
- https://github.com/Alexpux/Cygwin/pull/6
- https://github.com/Alexpux/Cygwin/pull/7
- https://github.com/Alexpux/Cygwin/pull/8

That last one needs most of the work: we have some patches for the POSIX <-> Windows path mangling that are required to pass Git's test suite (without which pass I will never even consider releasing a Git package). One of those patches, however, breaks MSYS2's expectations, and I have to rework that patch.

**Comment by Swoorup** [comment]
This is pretty cool. And after installing in my msys2 I have found no issues so far.

**Comment by dscho** [comment]
@Swoorup then you did not run Git's regression test suite... :wink:

**Comment by Swoorup** [comment]
i use quite a few git commands, rebase, status, commit, pull, revert, amend, merge. 

I am not really familiar about the git internals.
Should I actually stop using it?

**Comment by dscho** [comment]
@Swoorup you can do what you want, of course. The issues are subtle at best, anyway.

**Comment by dscho** [comment]
Unfortunately, [it has become clear that it would cost too much time and effort to get Git for Windows' patches into MSYS2](https://github.com/Alexpux/MSYS2-packages/pull/786#issuecomment-274512998).

On a happy note: this frees up a lot of time to work on making Git for Windows better! :smiley:

**Comment by mehrdadn** [comment]
Perhaps update the wiki to mention this won't be happening? Currently it says this is planned.

**Comment by dscho** [comment]
@mehrdadn where does it say that? Could you update that, too?

**Comment by mehrdadn** [comment]
@dscho I meant [here](/git-for-windows/git/wiki/Install-inside-MSYS2-proper). Sure, I'll try to change it.

**Comment by dscho** [comment]
@mehrdadn thanks!

**Comment by AlexandreMarkus** [comment]
> @fabio-porcedda This is already possible if you want. I'm doing this since a few months without problems.
> 
> I added instructions on the wiki: https://github.com/git-for-windows/git/wiki/Install-inside-MSYS2-proper

Working for me!!!

**Comment by 89z** [comment]
https://github.com/msys2/MINGW-packages/issues/10749

**Comment by rashil2000** [comment]
Hi! 

Now that both Cygwin version and Bash version are in sync with MSYS2 upstream (i.e. with the release of [Git for Windows 2.37.2](https://github.com/git-for-windows/git/releases/tag/v2.37.2.windows.1)), would it be worth revisiting this?

I'm aware that https://github.com/git-for-windows/git/wiki/Install-inside-MSYS2-proper exists, and I'm more than happy to use it, but it would be very convenient to have a fresh installation of MSYS2 and run just `pacman -S mingw-w64-x86_64-git` inside it like all other packages.

**Comment by dscho** [comment]
@rashil2000 @maehne @BasixKOR @pabranch @int-index @gphg I would like to encourage you to think long and hard about any way you could help bring this about. It is a huge task you are asking for.

**Comment by maehne** [comment]
@dscho: Could you give a summary, what tasks are open to make this happen? As this issue was created 7 years ago, I imagine some things have moved over this time. I still would like to see this, as I regularly use software packaged in MSYS2. I appreciate the frequent updates to Git for Windows (which I manage with `widget`), but it would be even more convenient if this could be done directly in MSYS using `pacman`.

Unfortunately, I cannot commit any resources to this issue, as I am already involved in several other projects leaving me no bandwidth. The only thing, I can offer, is testing.

**Comment by dscho** [comment]
> Could you give a summary, what tasks are open to make this happen?

Most of the adjustments needed to run Git smoothly are in [the fork of msys2-runtime](https://github.com/git-for-windows/msys2-runtime/) as well as in [the post-install script of the `git-extra` package](https://github.com/git-for-windows/build-extra/blob/HEAD/git-extra/git-extra.install.in).

> I appreciate the frequent updates to Git for Windows (which I manage with `widget`), but it would be even more convenient if this could be done directly in MSYS using `pacman`.

That convenience would come at the cost of an increased maintenance burden, and there are plenty of users who seem very comfortable to bow completely out of paying that cost.

> Unfortunately, I cannot commit any resources to this issue, as I am already involved in several other projects leaving me no bandwidth.

The problem is that everyone agrees with this stance, myself included: I cannot commit any resources to this issue. Some people find it reasonable to ask others to commit resources to this issue while they're unwilling/unable to do the same, but I am not in that group of people.

I even wonder why bother asking for what needs to be done, when that request is followed immediately by the statement that the requester won't do any of it. What outcome is expected other than to cause frustration all around?
